README.md
# china-mobile-phone-regexp

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![David deps][david-image]][david-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/china-mobile-phone-regexp.svg?style=flat-square
[npm-url]: https://npmjs.org/package/china-mobile-phone-regexp
[travis-image]: https://img.shields.io/travis/node-modules/china-mobile-phone-regexp.svg?style=flat-square
[travis-url]: https://travis-ci.org/node-modules/china-mobile-phone-regexp
[codecov-image]: https://img.shields.io/codecov/c/github/node-modules/china-mobile-phone-regexp.svg?style=flat-square
[codecov-url]: https://codecov.io/github/node-modules/china-mobile-phone-regexp?branch=master
[david-image]: https://img.shields.io/david/node-modules/china-mobile-phone-regexp.svg?style=flat-square
[david-url]: https://david-dm.org/node-modules/china-mobile-phone-regexp
[snyk-image]: https://snyk.io/test/npm/china-mobile-phone-regexp/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/china-mobile-phone-regexp
[download-image]: https://img.shields.io/npm/dm/china-mobile-phone-regexp.svg?style=flat-square
[download-url]: https://npmjs.org/package/china-mobile-phone-regexp

[The regular expressions](https://github.com/fengmk2/ChinaMobilePhoneNumberRegex) for China mobile phones.

## Install

```bash
$ npm i china-mobile-phone-regexp --save
```

## Usage

```js
const phone = require('china-mobile-phone-regexp');
// Is phone number?
console.log(phone.is('13800138000'));

// Is CHINA_MOBILE phone number?
console.log(phone.is('13800138000', phone.CHINA_MOBILE));

// Is CHINA_UNICOM phone number?
console.log(phone.is('13800138000', phone.CHINA_UNICOM));

// Is CHINA_TELECOM phone number?
console.log(phone.is('13800138000', phone.CHINA_TELECOM));
```

## Questions & Suggestions

Please open an issue [here](https://github.com/node-modules/china-mobile-phone-regexp/issues).

## License

[MIT](LICENSE)


History.md

1.0.0 / 2018-08-18
==================

**others**
  * [[`7a8e7ac`](http://github.com/node-modules/china-mobile-phone-regexp/commit/7a8e7ac482085768f9ef611c2f5a178b19865a16)] - test: add unit tests (fengmk2 <<fengmk2@gmail.com>>),fatal: No names found, cannot describe anything.



test/benchmark.js
'use strict';

const Benchmark = require('benchmark');

const suite = new Benchmark.Suite();

const lookahead = /^(?=\d{11}$)^1(?:3\d|4[4-9]|5[0-35-9]|6[67]|7[013-8]|8\d|9\d)\d{8}$/;
const normal = /^1(?:3\d|4[4-9]|5[0-35-9]|6[67]|7[013-8]|8\d|9\d)\d{8}$/;

console.log(normal.test('18668079069'), normal.test('1866807906a'));
console.log(lookahead.test('18668079069'), lookahead.test('1866807906a'));

// add tests
suite.add('normal', function() {
  normal.test('18668079069');
  normal.test('1866807906a');
  normal.test('1');
  normal.test('a');
})
  .add('enable lookahead', function() {
    lookahead.test('18668079069');
    lookahead.test('1866807906a');
    normal.test('1');
    normal.test('a');
  })
// add listeners
  .on('cycle', function(event) {
    console.log(String(event.target));
  })
  .on('complete', function() {
    console.log('Fastest is ' + this.filter('fastest').map('name'));
  })
// run async
  .run({ async: true });


test/index.test.js
'use strict';

const assert = require('assert');
const utility = require('utility');
const phone = require('..');

describe('test/index.test.js', () => {
  describe('ALL', () => {
    phone.CHINA_MOBILE_PREFIXS.forEach(prefix => genIT(prefix));
    phone.CHINA_UNICOM_PREFIXS.forEach(prefix => genIT(prefix));
    phone.CHINA_TELECOM_PREFIXS.forEach(prefix => genIT(prefix));
  });

  describe('CHINA_MOBILE', () => {
    it('should work with country code', () => {
      assert(phone.is('13800000000', phone.CHINA_MOBILE));
      assert(phone.is(genNumber('138'), phone.CHINA_MOBILE));
      assert(phone.is('08613800000000', phone.CHINA_MOBILE));
      assert(phone.is('008613800000000', phone.CHINA_MOBILE));
      assert(phone.is('+8613800000000', phone.CHINA_MOBILE));
      assert(phone.is('+08613800000000', phone.CHINA_MOBILE));
      assert(phone.is('+008613800000000', phone.CHINA_MOBILE));

      assert(!phone.is('+008813800000000', phone.CHINA_MOBILE));
      assert(!phone.is('+008618600000000', phone.CHINA_MOBILE));
    });

    phone.CHINA_MOBILE_PREFIXS.forEach(prefix => genIT(prefix, phone.CHINA_MOBILE));
  });

  describe('CHINA_UNICOM', () => {
    phone.CHINA_UNICOM_PREFIXS.forEach(prefix => genIT(prefix, phone.CHINA_UNICOM));
  });

  describe('CHINA_TELECOM', () => {
    phone.CHINA_TELECOM_PREFIXS.forEach(prefix => genIT(prefix, phone.CHINA_TELECOM));
  });

  describe('MVNO', () => {
    phone.MVNO_PREFIXS.forEach(prefix => genIT(prefix, phone.MVNO));
  });

  describe('SATELLITE', () => {
    phone.SATELLITE_PREFIXS.forEach(prefix => genIT(prefix, phone.SATELLITE));
  });

  describe('DATA', () => {
    phone.DATA_PREFIXS.forEach(prefix => genIT(prefix, phone.DATA));
  });
});

function genIT(prefix, regexp) {
  it(`should match ${prefix}`, () => {
    const splits = prefix.split('/');
    const first = splits[0];
    const len = first.length - 1;
    for (let i = 0; i < splits.length; i++) {
      let p = splits[i];
      if (i > 0) {
        p = `${first.substring(0, len)}${p}`;
      }
      console.log('%s => %s', p, genNumber(p));
      genNumbers(p).forEach(n => assert(phone.is(n, regexp)));

      if (regexp === phone.CHINA_MOBILE) {
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.CHINA_UNICOM)));
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.CHINA_TELECOM)));
      } else if (regexp === phone.CHINA_UNICOM) {
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.CHINA_MOBILE)));
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.CHINA_TELECOM)));
      } else if (regexp === phone.CHINA_TELECOM) {
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.CHINA_MOBILE)));
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.CHINA_UNICOM)));
      }

      if (regexp === phone.MVNO) {
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.SATELLITE)));
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.DATA)));
      } else if (regexp === phone.SATELLITE) {
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.MVNO)));
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.DATA)));
      } else if (regexp === phone.DATA) {
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.MVNO)));
        genNumbers(p).forEach(n => assert(!phone.is(n, phone.SATELLITE)));
      }
    }
  });
}

function genNumbers(prefix, count) {
  count = count || 10000;
  const numbers = [];
  for (let i = 0; i < count; i++) {
    numbers.push(genNumber(prefix));
  }
  return numbers;
}

function genNumber(prefix) {
  prefix = prefix || '1';
  return prefix + utility.randomString(11 - prefix.length, '0123456789');
}


index.js
'use strict';

// Match all
// 全部
const ALL = /^1(?:3\d|4[04-9]|41(?=[1-9])|5[0-35-9]|6[67]|7[013-8]|8\d|9\d)\d{8}$/;

// MVNO: 167, 170/1
// 虚拟运营商
const MVNO_PREFIXS = [
  '167',
  '170/1',
];
const MVNO = /^1(?:67|7[01])\d{8}$/;
const EXCLUDE_MVNO = /^1(?:3\d|4[4-9]|5[0-35-9]|66|7[3-8]|8\d|9\d)\d{8}$/;

// SATELLITE: 1349, 17400/1/2/3/4/5, 1749
// 卫星电话
const SATELLITE_PREFIXS = [
  '1349',
  '17400/1/2/3/4/5',
  '1749',
];
const SATELLITE = /^1(?:349|749|740(?=[0-5]))\d{7}$/;
const EXCLUDE_SATELLITE = /^1(?:3(?!49)\d|4[4-9]|5[0-35-9]|6[67]|7[0135-8]|8\d|9\d)\d{8}$/;

// DATA: 144/5/6/7/8/9
// 数据卡
const DATA_PREFIXS = [
  '144/5/6/7/8/9',
];
const DATA = /^1(?:4[4-9])\d{8}$/;
const EXCLUDE_DATA = /^1(?:3\d|5[0-35-9]|6[67]|7[013-8]|8\d|9\d)\d{8}$/;

const CHINA_MOBILE_PREFIXS = [
  '1340/1/2/3/4/5/6/7/8',
  '135/6/7/8/9',
  // 14410~14499 + 1440 => 144
  '144/7/8',
  '150/1/2/7/8/9',
  '1703/5/6',
  '178',
  '182/3/4/7/8',
  '198',
];
const CHINA_MOBILE = /^1(?:34(?=[0-8])|3[5-9]|4[478]|5[012789]|70(?=[356])|78|8[23478]|98)\d{8}$/;

const CHINA_UNICOM_PREFIXS = [
  '130/1/2',
  // 14000~14099 => 140
  '140',
  '145/6',
  '155/6',
  '167',
  '1704/7/8/9',
  '171/5/6',
  '185/6',
];
const CHINA_UNICOM = /^1(?:3[012]|4[056]|5[56]|67|70(?=[4789])|7[156]|8[56])\d{8}$/;

const CHINA_TELECOM_PREFIXS = [
  '133',
  '1349',
  // 14110~14199 => 1411/2/3/4/5/6/7/8/9
  '1411/2/3/4/5/6/7/8/9',
  '149',
  '153',
  '1700/1/2',
  '173/7',
  '17400/1/2/3/4/5',
  '180/1/9',
  '191/9',
];
const CHINA_TELECOM = /^1(?:33|34(?=9)|41(?=[1-9])|49|53|70(?=[012])|7[37]|74(?=0[0-5])|8[019]|9[19])\d{8}$/;

// +86
// 086
// 0086
// +0086
// +086
const COUNTRY_CODE = /^(\+?0{0,2}86)/;

module.exports = {
  ALL,
  CHINA_MOBILE,
  CHINA_UNICOM,
  CHINA_TELECOM,
  MVNO,
  EXCLUDE_MVNO,
  SATELLITE,
  EXCLUDE_SATELLITE,
  DATA,
  EXCLUDE_DATA,
  COUNTRY_CODE,

  MVNO_PREFIXS,
  SATELLITE_PREFIXS,
  DATA_PREFIXS,
  CHINA_MOBILE_PREFIXS,
  CHINA_UNICOM_PREFIXS,
  CHINA_TELECOM_PREFIXS,

  is(number, regexp) {
    regexp = regexp || ALL;
    if (number.length > 11) {
      const m = COUNTRY_CODE.exec(number);
      if (m) {
        number = number.substring(m[1].length);
      }
    }

    return number.length === 11 && regexp.test(number);
  },
};
